KAFKA-7518: FutureRecordMetadata.get deadline calculation fix

akatona84 opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

akatona84 committed:
KAFKA-7518: FutureRecordMetadata.get deadline calculation from timeout is not using timeunit

akatona84 commented:
@viktorsomogyi @asasvari @andrasbeni @smurakozi pls review

omkreddy commented:
@akatona84 Good catch. maybe we can add or modify existing test to verify the change.

ijuma commented:
Great catch.

akatona84 committed:
Adding test which fails with original deadline calculation

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadataTest.java:
```diff
@@ -0,0 +1,55 @@
+package org.apache.kafka.clients.producer.internals;
```
Can you please add the copyright notice at the top?

ijuma commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadata.java:
```diff
@@ -68,12 +68,13 @@ public RecordMetadata get() throws InterruptedException, ExecutionException {
         boolean occurred = this.result.await(timeout, unit);
         if (nextRecordMetadata != null)
             return nextRecordMetadata.get(deadline - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
         if (!occurred)
```
Seems like we should check if it's occurred before calling `nextRecordMetadata.get`.

akatona84 replied:
True. Timeout error throwing is moved up a bit

akatona84 committed:
Checking if actual data get even happened in FutureRecordMetadata.get
before calling nextRecordMetadata.get

akatona84 committed:
Added copyright

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadataTest.java:
```diff
@@ -0,0 +1,71 @@
+        // nothing to do
+      }
+    });
+    thread.start();
```
Instead of adding a thread and a sleep, which makes tests slower, we could simply verify that the right value is passed to `ProduceRequestResult.await` and `nextRecordMetadata.get` by using Mockito. What do you think?

akatona84 replied:
I wanted to mock first then realised that they are final classes... so these classes cannot be mocked, at least not with the current version of mockito.

ijuma replied:
They are internal classes, so we can remove `final` if that helps with testing.

akatona84 replied:
Updated tests, actually Mockito can mock final classes, just had to add an extra test resource to enable it. I wanted to avoid removing final from classed just for testing purposes.

akatona84 committed:
Added common.utils.Time to FutureRecordMetadata, tests modify/refactor

- Time is set by constructor, added to occurrences
- test is modified to verify it is called with proper timeouts and units
- eliminated threaded timeout handling via MockTime (faster test)

akatona84 commented:
retest this please

akatona84 commented:
@omkreddy @ijuma what do you think about the current state of this change?
- I was able to eliminate starting separate thread in my test and in another by introducing the Time interface in this internal class as well
- FutureRecordMetadata changed as suggested (failing fast)

But I don't know what the problem is with junit tests, they keep failing (not because of my change)

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/RecordSendTest.java:
```diff
@@ -16,28 +16,30 @@
     private final TopicPartition topicPartition = new TopicPartition("test", 0);
     private final long baseOffset = 45;
     private final long relOffset = 5;
+    private Time time = new MockTime();
```
Do we need this field at all? Can we just pass the time in each method?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/RecordSendTest.java:
```diff
@@ -75,24 +77,17 @@ public void testError() throws Exception {
 
     /* create a new request result that will be completed after the given timeout */
-    public ProduceRequestResult asyncRequest(final long baseOffset, final RuntimeException error, final long timeout) {
+    private ProduceRequestResult produceRequestResult(final long baseOffset, final RuntimeException error) {
```
This comment needs to be updated since we are not passing a `timeout`. What's the intent behind this change?

akatona84 replied:
Here the method started a separate thread and played with Thread.sleep which makes tests slower and flakier.
I had to modify the test class when introduced Time interface in FutureRecordMetadata class, and it made me to change this tests too to be faster/more efficient.

ijuma replied:
It seems like you're changing the semantics of the tests though. It may be best to leave them as they were.

ijuma replied:
It seems like you're changing the semantics of the tests though. It may be best to leave them as they were.

akatona84 replied:
ok, i c. 
reverted

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadataTest.java:
```diff
@@ -0,0 +1,88 @@
+
+public class FutureRecordMetadataTest {
+
+    private MockTime time = new MockTime();
```
Nit: can this be `final`?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadataTest.java:
```diff
@@ -0,0 +1,88 @@
+
+    @Test
+    public void testFutureGetWithSeconds() throws ExecutionException, InterruptedException, TimeoutException {
+        // given
```
Nit: we don't normally use this type of comment.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/FutureRecordMetadataTest.java:
```diff
@@ -0,0 +1,88 @@
+        verify(chainedProduceRequestResult).await(1000L, TimeUnit.MILLISECONDS);
+    }
+
+    private FutureRecordMetadata getFutureRecordMetadata(ProduceRequestResult produceRequestResult) {
```
Nit: `futureRecordMetadata`, no need for `get`.

ijuma commented on clients/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Why do we need this?

akatona84 replied:
to be able to mock final classes with mockito

ijuma replied:
This is not worth it: https://github.com/mockito/mockito/issues/1142

As I said, we can remove `final` from internal classes if needed for testing.

akatona84 replied:
good point, I removed. just wanted to avoid prod code change just for tests

ijuma reviewed (commented):
Thanks for the update. Left a few comments. Can you please rebase against trunk?

akatona84 committed:
review related changes

akatona84 committed:
Reverting RecordSendTest changes

akatona84 committed:
Removing final to be able to mock, deleting mockito-extensions

ijuma reviewed (approved):
Thanks for the PR, LGTM.

ijuma commented:
Merged to trunk and 2.1 branches.
